pkg/operator/controller/instance_controller_test.go

```diff
@@ -0,0 +1,319 @@
+	ingressID := types.NamespacedName{Name: controller.IngressName, Namespace: controller.InstanceNamespace}
+
+	t.Run("Create ClusterLink deployment", func(t *testing.T) {
+		// Create ClusterLink namespaces
+		opearatorNamespace := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: controller.OperatorNameSpace,
+			},
+		}
+		Expect(k8sClient.Create(ctx, opearatorNamespace)).Should(Succeed())
+
+		systemNamespace := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: controller.InstanceNamespace,
+			},
+		}
+		Expect(k8sClient.Create(ctx, systemNamespace)).Should(Succeed())
+
+		// Create ClusterLink deployment
+		Expect(k8sClient.Create(ctx, &cl)).Should(Succeed())
+
+		// Check Controlplane resources
+		for _, r := range cpResource {
+			checkResourceCreated(cpID, r)
+		}
+
+		// Check controlplane fields
+		cp := &appsv1.Deployment{}
+		getResource(cpID, cp)
+		cpImage := "ghcr.io/clusterlink-net/" + controller.ControlPlaneName + ":latest"
+		Expect(cp.Spec.Template.Spec.Containers[0].Image).Should(Equal(cpImage))
+		Expect(cp.Spec.Template.Spec.Containers[0].Args[1]).Should(Equal("info"))
+
+		// Check Dataplane resources
+		for _, r := range dpResource {
+			checkResourceCreated(dpID, r)
+		}
+
+		// Check Dataplane fields
+		dp := &appsv1.Deployment{}
+		getResource(dpID, dp)
+		envoyImage := "ghcr.io/clusterlink-net/" + controller.DataPlaneName + ":latest"
+		Expect(dp.Spec.Template.Spec.Containers[0].Image).Should(Equal(envoyImage))
+		Expect(*dp.Spec.Replicas).Should(Equal(int32(1)))
+		Expect(dp.Spec.Template.Spec.Containers[0].Args[1]).Should(Equal("info"))
+
+		// Check ingress resource not exist
+		assert.True(t, checkResourceNotExist(ingressID, &corev1.Service{}))
+	})
+
+	t.Run("Update ClusterLink deployment", func(t *testing.T) {
+		svc := &corev1.Service{}
+		// Update ingress resource to LoadBalancer type
+		getResource(types.NamespacedName{Name: ClusterLinkName, Namespace: controller.OperatorNameSpace}, &cl)
+		cl.Spec.Ingress.Type = clusterlink.IngressTypeLoadBalancer
+		Expect(k8sClient.Update(ctx, &cl)).Should(Succeed())
+		checkResourceCreated(ingressID, svc)
+		Expect(svc.Spec.Type).Should(Equal(corev1.ServiceTypeLoadBalancer))
+
+		// Update ingress resource to NodPport type
+		Expect(k8sClient.Delete(ctx, svc)).Should(Succeed())
+		getResource(types.NamespacedName{Name: ClusterLinkName, Namespace: controller.OperatorNameSpace}, &cl)
+		cl.Spec.Ingress.Type = clusterlink.IngressTypeNodePort
+		cl.Spec.Ingress.Port = 30444
+		Expect(k8sClient.Update(ctx, &cl)).Should(Succeed())
+		checkResourceCreated(ingressID, svc)
+		Expect(svc.Spec.Type).Should(Equal(corev1.ServiceTypeNodePort))
+		Expect(svc.Spec.Ports[0].NodePort).Should(Equal(int32(30444)))
+
+		// Update dataplane and controlpane
+		goReplicas := 2
+		loglevel := "debug"
+		containerRegistry := "quay.com"
+		imageTag := "v1.0.1"
+		cp := &appsv1.Deployment{}
+		dp := &appsv1.Deployment{}
+
+		getResource(cpID, cp)
```

Comment:
shouldn`t deletion by done by the operator in response to changes in the Spec? Why do you need to manually delete (which I would expect the controller to be notified on and recreate the resource immediately...)?